### PR TITLE
FIX: Redis to handle event metadata.

### PR DESCRIPTION
**Pull request by artursouza**
# Description

Redis does not handle event metadata at all. We had a P0 issue in runtime that was discovered because of Pulsar pubsub and it could not be reproduced with Redis because of inconsistency of handling event metadata - in this case, Redis simply drops it. This PR brings the Redis component (stable) at the same level of compatibility we should have across beta and stable pubsub components (IMO) regarding event metadata handling. This is not a P0 (IMO), it will just block the E2E test to reproduce the P0 bug we will need to fix.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Related to the following issues:
 - P0 in runtime that fixes conflict of event metadata for pubsub: https://github.com/dapr/dapr/issues/7413
 - non-P0 issue to discuss consistency across pubsub for event metadata handling: https://github.com/dapr/components-contrib/issues/3321

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

RELEASE_NOTE: **FIXED** Redis pubsub now handles event metadata.

### CODE REVIEW

File: pubsub/redis/redis.go
Comment (artursouza) [inline pubsub/redis/redis.go:202]: This seems to be better than getting the message stuck because metadata could not be handled.

**Review by yaron2 — APPROVED** [review]

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
Since there's no linked issue, can you please explain what this PR does and why it's needed?

We are past the code freeze phase, so this should probably be merged when the 1.14 merge window opens, unless there's a P0 issue attached to it.

**Comment by artursouza** [comment]
I can create a specific issue if this needs further discussion. I created two issues that this PR is part of. Basically, bringing having consistency across pubsub components regarding handling of pubsub metadata.

This will only block our integration or E2E tests to reproduce the bug I am fixing for 1.13 - which is a P0 as it makes pubsub unusable for some components with gRPC subscriber.

**Comment by artursouza** [comment]
> Since there's no linked issue, can you please explain what this PR does and why it's needed?
> 
> We are past the code freeze phase, so this should probably be merged when the 1.14 merge window opens, unless there's a P0 issue attached to it.

See. See the updated description and my comment. This fix can wait until 1.14 but we will not have E2E tests to validate the hotfix for 1.13 - which I can do manually. Of course, we can add yet-another-pubsub to E2E tests, but I don't think it is the right path since this is a basic inconsistency in pubsub that should be fixed (not P0).

**Comment by ItalyPaleAle** [comment]
@artursouza my .02:

- If this is needed primarily to unblock an E2E test in runtime, I'd prefer if it were added to the in-memory pubsub component for 1.13
- If we agree that all components should support metadata (or at least expose a feature flag for those who do support that or not), then we should have a conformance test to validate this, or the implementation could end up being "patchwork-y"

I'd prefer avoiding adding new features to a "tier 1" component (with a large user base like Redis) when we're already in endgame. But I can be open if other maintainers agree this is the best path forward.

**Comment by artursouza** [comment]
> @artursouza my .02:
> 
> 
> 
> - If this is needed primarily to unblock an E2E test in runtime, I'd prefer if it were added to the in-memory pubsub component for 1.13
> 
> - If we agree that all components should support metadata (or at least expose a feature flag for those who do support that or not), then we should have a conformance test to validate this, or the implementation could end up being "patchwork-y"
> 
> 
> 
> I'd prefer avoiding adding new features to a "tier 1" component (with a large user base like Redis) when we're already in endgame. But I can be open if other maintainers agree this is the best path forward.

In-memory pubsub does not work for e2e because the subscriber is not the same app as the publisher. While the test can exist with another pubsub, I found it odd for such a basic feature not be supported by all beta and stable pubsub. If this behavior is not standard, then we should add a new column to the feature set of pubsub components. As you see, I opened an issue to discuss the expected behavior for pubsubs and see what others think.

In the worst case, I can add this feature behind a component metadata flag for those that need it.

File: pubsub/redis/redis.go
Comment (berndverst) [inline pubsub/redis/redis.go:186]: This change is not necessary. It exists the way it is by design because this helper method does not operate on the struct at all. It is static.

Comment (berndverst) [inline pubsub/redis/redis.go:186]:
```suggestion
func createRedisMessageWrapper(ctx context.Context, stream string, handler pubsub.Handler, msg rediscomponent.RedisXMessage) redisMessageWrapper {
```

Comment (berndverst) [reply pubsub/redis/redis.go:186]: Ah, I see you did it because you wanted access to the logger. Ok that's fine then

**Review by berndverst — CHANGES_REQUESTED** [review]
I will approve this, but I really think you should add conformance tests.

I will not block on conformance tests, but that is the right thing to do.

My only blocking comment is to please prefix the error message with the component name as I have suggested. This helps with troubleshooting or grepping for similar logs (to see prevalence of an issue etc etc).

Beyond that I am fine merging this.

**Review by ItalyPaleAle — APPROVED** [review]
Given that @berndverst  and @yaron2 both agree to include this, I'm fine too. Thanks for your patience :)

**Comment by berndverst** [comment]
I took the liberty to apply my suggested log prefix. LGTM now.

**Review by berndverst — APPROVED** [review]

**Pull request merged**
